qiskit/dagcircuit/collect_blocks.py

```diff
@@ -0,0 +1,255 @@
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""Various ways to divide a DAG into blocks of nodes, to split blocks of nodes
+into smaller sub-blocks, and to consolidate blocks."""
+
+from qiskit.circuit import QuantumCircuit, CircuitInstruction
+from . import DAGOpNode, DAGCircuit, DAGDependency
+from .exceptions import DAGCircuitError
+
+
+class BlockCollector:
+    """This class implements various strategies of dividing a DAG (direct acyclic graph)
+    into blocks of nodes that satisfy certain criteria. It works both with the
+    :class:`~qiskit.dagcircuit.DAGCircuit` and
+    :class:`~qiskit.dagcircuit.DAGDependency` representations of a DAG, where
+    DagDependency takes into account commutativity between nodes.
+
+    Collecting nodes from DAGDependency generally leads to more optimal results, but is
+    slower, as it requires to construct a DAGDependency beforehand. Thus, DAGCircuit should
+    be used with lower transpiler settings, and DAGDependency should be used with higher
+    transpiler settings.
+
+    In general, there are multiple ways to collect maximal blocks. The approaches used
+    here are of the form 'starting from the input nodes of a DAG, greedily collect
+    the largest block of nodes that match certain criteria'. For additional details,
+    see https://github.com/Qiskit/qiskit-terra/issues/5775.
+    """
+
+    def __init__(self, dag):
+        """
+        Args:
+            dag (Union[DAGCircuit, DAGDependency]): The input DAG.
+
+        Raises:
+            DAGCircuitError: the input object is not a DAG.
+        """
+
+        self.dag = dag
+
+        if isinstance(dag, DAGCircuit):
+            self.is_dag_dependency = False
+
+        elif isinstance(dag, DAGDependency):
+            self.is_dag_dependency = True
+
+        else:
+            raise DAGCircuitError("not a DAG.")
+
+        # For an efficient implementation, we compute and keep updating the in_degree
+        # for every node, that is the number of node's immediate predecessors.
+        # A node is a leaf (aka input) node iff its in_degree is 0.
+        # When a node is (marked as) collected, the in_degrees of its immediate
+        # successors are updated by subtracting 1.
+        # Additionally, pending_nodes explicitly keeps the list of nodes with in_degree 0.
+        self.pending_nodes = []
+        self.in_degree = dict()
+        for node in self._op_nodes():
+            deg = len(self._direct_preds(node))
+            self.in_degree[node] = deg
+            if deg == 0:
+                self.pending_nodes.append(node)
```

Comment:
With this code in `__init__, is it possible to run the collection methods more than once? e.g.
```
collector = BlockCollector(dag)
first_blocks = collector.collect_all_matching_blocks(filter_1)
second_blocks = collector.collect_all_matching_blocks(filter_2)
```

Reply:
Good catch! Currently we would not be able to do the above, I will change that.

Reply:
I have realized that there are two somewhat different ways to interpret the code above. 

Let's suppose that `filter_1` looks for linear gates (i. e. CX and SWAP), while `filter_2` looks for Clifford gates (i. e. CX, SWAP, S, H, etc.).

My default interpretation is that some gates may belong both to `first_blocks` and to `second_blocks`. This could be useful when we have a quantum circuit and we are unsure which of the synthesis methods to apply (i.e. should we collect linear gates and resynthesize linear functions?, or should we collect clifford gates and resynthesize cliffords?), and we might make this choice by comparing sizes of blocks collected by both methods (for instance, if collecting clifford gates does not seem to offer much advantage over collecting linear gates, then we should not do clifford synthesis; though this reasoning may actually be bad since in some cases linear functions can be synthesized better by allowing the synthesis algorithm to  use clifford gates). However, this does not allow us to collapse `first_blocks` first and `second_blocks` second, that is the `second_blocks` may become invalidated when some of the `first_blocks` are collapsed. 

An alternative interpretation is that `second_blocks` should only consider gates not collected in `first_blocks`. I like this less, especially that we can get the combined result by calling the main `CollapseChains` pass twice:
```
dag1 = CollectLinearFunctions().run(dag)
dag2 = CollectCliffords().run(dag1)
```
We could also think how to further extend the `CollapseChains` pass to take multiple collect/collapse functions. 

If there are no strong objections, I will proceed with the first interpretation. I don't think this will give any performance improvement (as we do not copy the underlying DAG) but it will make the code a bit easier to use.

Reply:
Update: this is handled in [d5f330e](https://github.com/Qiskit/qiskit-terra/pull/8907/commits/d5f330eb00a6a64a8ba7a626e383a584322c7819), including tests to show that this works.